### ISSUE TITLE
kube-apiserver v1.28.3 panics with fatal error: concurrent map iteration and map write

### ISSUE THREAD

**Issue by songminglong** [kubernetes/kubernetes#122940]
### What happened?

I am doing v1.28.3 kube-apiserver testing in the test environment, and I often encounter kube-apiserver panic, especially when kube-apiserver call webhook fails.

The following are the captured panic logs:
```
Jan 18 08:31:00 VMS71884 kube-apiserver[913778]: E0118 08:31:00.777233  913778 dispatcher.go:213] failed calling webhook "validation.istio.io": failed to call webhook: Post "https://istiod.istio-system.svc:443/validate?timeout=30s": service "istiod" not found
Jan 18 08:31:01 VMS71884 kube-apiserver[913778]: E0118 08:31:01.703058  913778 dispatcher.go:213] failed calling webhook "validation.istio.io": failed to call webhook: Post "https://istiod.istio-system.svc:443/validate?timeout=30s": service "istiod" not found
Jan 18 08:31:06 VMS71884 kube-apiserver[913778]: fatal error: concurrent map iteration and map write
Jan 18 08:31:06 VMS71884 kube-apiserver[913778]: goroutine 177317973 [running]:
Jan 18 08:31:06 VMS71884 kube-apiserver[913778]: golang.org/x/net/http2.cloneHeader(0xc0aa793b00)
Jan 18 08:31:06 VMS71884 kube-apiserver[913778]: vendor/golang.org/x/net/http2/server.go:2944 +0xb0
Jan 18 08:31:06 VMS71884 kube-apiserver[913778]: golang.org/x/net/http2.(*responseWriterState).writeHeader(0xc1257f9200, 0x1f8)
Jan 18 08:31:06 VMS71884 kube-apiserver[913778]: vendor/golang.org/x/net/http2/server.go:2938 +0x3df
Jan 18 08:31:06 VMS71884 kube-apiserver[913778]: golang.org/x/net/http2.(*responseWriter).WriteHeader(0xc140fdbf08, 0x1f8)
Jan 18 08:31:06 VMS71884 kube-apiserver[913778]: vendor/golang.org/x/net/http2/server.go:2900 +0x39
Jan 18 08:31:06 VMS71884 kube-apiserver[913778]: go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp.(*respWriterWrapper).WriteHeader(0xc12261ff20, 0x1f8)
Jan 18 08:31:06 VMS71884 kube-apiserver[913778]: vendor/go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp/wrap.go:94 +0x63
Jan 18 08:31:06 VMS71884 kube-apiserver[913778]: github.com/felixge/httpsnoop.(*rw).WriteHeader(0xc0e082c640, 0x1f8)
Jan 18 08:31:06 VMS71884 kube-apiserver[913778]: vendor/github.com/felixge/httpsnoop/wrap_generated_gteq_1.8.go:372 +0xe3
Jan 18 08:31:06 VMS71884 kube-apiserver[913778]: k8s.io/apiserver/pkg/server/httplog.(*respLogger).WriteHeader(0xc0e19e8370, 0x1f8)
Jan 18 08:31:06 VMS71884 kube-apiserver[913778]: vendor/k8s.io/apiserver/pkg/server/httplog/httplog.go:301 +0x47
Jan 18 08:31:06 VMS71884 kube-apiserver[913778]: k8s.io/apiserver/pkg/server/filters.(*baseTimeoutWriter).timeout(0xc0aa793f50, 0xc0e13f3cc0)
Jan 18 08:31:06 VMS71884 kube-apiserver[913778]: vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:258 +0x11d
Jan 18 08:31:06 VMS71884 kube-apiserver[913778]: k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP(0xc00964eb58, {0x6adf0f0, 0xc1095cc860}, 0xc0e0831b00)
Jan 18 08:31:06 VMS71884 kube-apiserver[913778]: vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:149 +0x555
Jan 18 08:31:06 VMS71884 kube-apiserver[913778]: k8s.io/apiserver/pkg/endpoints/filters.withRequestDeadline.func1({0x6adf0f0, 0xc1095cc820}, 0xc0e0831b00)
Jan 18 08:31:06 VMS71884 kube-apiserver[913778]: vendor/k8s.io/apiserver/pkg/endpoints/filters/request_deadline.go:100 +0x73e
Jan 18 08:31:06 VMS71884 kube-apiserver[913778]: net/http.HandlerFunc.ServeHTTP(0xc0165d3a40, {0x6adf0f0, 0xc1095cc820}, 0xc0e0831a00)
Jan 18 08:31:06 VMS71884 kube-apiserver[913778]: /usr/local/go/src/net/http/server.go:2122 +0x43
Jan 18 08:31:06 VMS71884 kube-apiserver[913778]: k8s.io/apiserver/pkg/server/filters.withWaitGroup.func1({0x6adf0f0, 0xc1095cc820}, 0xc0e0831a00)
Jan 18 08:31:06 VMS71884 kube-apiserver[913778]: vendor/k8s.io/apiserver/pkg/server/filters/waitgroup.go:86 +0x2fe
Jan 18 08:31:06 VMS71884 kube-apiserver[913778]: net/http.HandlerFunc.ServeHTTP(0xc0165e02c0, {0x6adf0f0, 0xc1095cc820}, 0xc0e0831a00)
Jan 18 08:31:06 VMS71884 kube-apiserver[913778]: /usr/local/go/src/net/http/server.go:2122 +0x43
Jan 18 08:31:06 VMS71884 kube-apiserver[913778]: k8s.io/apiserver/pkg/server/filters.(*goaway).ServeHTTP(0xc0165b15a0, {0x6adf0f0, 0xc1095cc820}, 0xc0e0831a00)





Jan 07 13:58:37 VMS71884 kube-apiserver[78343]: E0107 13:58:37.697995   78343 dispatcher.go:214] failed calling webhook "chaosblade-operator.chaosblade.svc": failed to call webhook: Post "https://chaosblade-webhook-server.chaosblade.svc:443/mutating-pods?timeout=10s": dial tcp 10.7.153.222:443: connect: connection refused
Jan 07 13:58:37 VMS71884 kube-apiserver[78343]: fatal error: concurrent map iteration and map write
Jan 07 13:58:37 VMS71884 kube-apiserver[78343]: goroutine 589955230 [running]:
Jan 07 13:58:37 VMS71884 kube-apiserver[78343]: golang.org/x/net/http2.(*responseWriterState).promoteUndeclaredTrailers(0xc0bb23a780)
Jan 07 13:58:37 VMS71884 kube-apiserver[78343]: vendor/golang.org/x/net/http2/server.go:2750 +0x1ca
Jan 07 13:58:37 VMS71884 kube-apiserver[78343]: golang.org/x/net/http2.(*responseWriterState).writeChunk(0xc0bb23a780, {0xc0197e1000, 0x98, 0x1000})
Jan 07 13:58:37 VMS71884 kube-apiserver[78343]: vendor/golang.org/x/net/http2/server.go:2611 +0xa6
Jan 07 13:58:37 VMS71884 kube-apiserver[78343]: golang.org/x/net/http2.chunkWriter.Write({0xc0bb23a780}, {0xc0197e1000, 0x98, 0x1000})
Jan 07 13:58:37 VMS71884 kube-apiserver[78343]: vendor/golang.org/x/net/http2/server.go:2564 +0x6c
Jan 07 13:58:37 VMS71884 kube-apiserver[78343]: bufio.(*Writer).Flush(0xc1669b2f00)
Jan 07 13:58:37 VMS71884 kube-apiserver[78343]: /usr/local/go/src/bufio/bufio.go:628 +0x11f
Jan 07 13:58:37 VMS71884 kube-apiserver[78343]: golang.org/x/net/http2.(*responseWriter).FlushError(0xc0e2a32038)
Jan 07 13:58:37 VMS71884 kube-apiserver[78343]: vendor/golang.org/x/net/http2/server.go:2829 +0x8c
Jan 07 13:58:37 VMS71884 kube-apiserver[78343]: golang.org/x/net/http2.(*responseWriter).Flush(0xc0e2a32038)
Jan 07 13:58:37 VMS71884 kube-apiserver[78343]: vendor/golang.org/x/net/http2/server.go:2819 +0x1e
Jan 07 13:58:37 VMS71884 kube-apiserver[78343]: golang.org/x/net/http2.(*responseWriter).handlerDone(0xc0e2a32038)
Jan 07 13:58:37 VMS71884 kube-apiserver[78343]: vendor/golang.org/x/net/http2/server.go:2997 +0x46
Jan 07 13:58:37 VMS71884 kube-apiserver[78343]: golang.org/x/net/http2.(*serverConn).runHandler.func1()
Jan 07 13:58:37 VMS71884 kube-apiserver[78343]: vendor/golang.org/x/net/http2/server.go:2366 +0x445
Jan 07 13:58:37 VMS71884 kube-apiserver[78343]: golang.org/x/net/http2.(*serverConn).runHandler(0xc103848680, 0xc0e2a32038, 0xc09a809b00, 0xc1b2789f68)
Jan 07 13:58:37 VMS71884 kube-apiserver[78343]: vendor/golang.org/x/net/http2/server.go:2370 +0x1ec
Jan 07 13:58:37 VMS71884 kube-apiserver[78343]: created by golang.org/x/net/http2.(*serverConn).scheduleHandler
Jan 07 13:58:37 VMS71884 kube-apiserver[78343]: vendor/golang.org/x/net/http2/server.go:2303 +0x1c9
Jan 07 13:58:37 VMS71884 kube-apiserver[78343]: goroutine 1 [chan receive, 7572 minutes]:
Jan 07 13:58:37 VMS71884 kube-apiserver[78343]: k8s.io/apiserver/pkg/server.preparedGenericAPIServer.Run({0xc00551a300}, 0xc0004601e0)
Jan 07 13:58:37 VMS71884 kube-apiserver[78343]: vendor/k8s.io/apiserver/pkg/server/genericapiserver.go:674 +0xdba
Jan 07 13:58:37 VMS71884 kube-apiserver[78343]: k8s.io/kube-aggregator/pkg/apiserver.preparedAPIAggregator.Run({0xc0049d54a0, {0x6aba220, 0xc00551a300}}, 0xc0004601e0)
Jan 07 13:58:37 VMS71884 kube-apiserver[78343]: vendor/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:467 +0x45
Jan 07 13:58:37 VMS71884 kube-apiserver[78343]: k8s.io/kubernetes/cmd/kube-apiserver/app.Run({0xc000dda000}, 0xc0004601e0)
Jan 07 13:58:38 VMS71884 kube-apiserver[78343]: cmd/kube-apiserver/app/server.go:188 +0x711
Jan 07 13:58:38 VMS71884 kube-apiserver[78343]: k8s.io/kubernetes/cmd/kube-apiserver/app.NewAPIServerCommand.func2(0xc000221800, {0xc000004900, 0x0, 0x2e})
Jan 07 13:58:38 VMS71884 kube-apiserver[78343]: cmd/kube-apiserver/app/server.go:136 +0x34e

```

and:

```
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: E0124 06:10:56.887453   86280 watch.go:287] unable to encode watch object *v1.WatchEvent: client disconnected (&streaming.encoderWithAllocator{writer:responsewriter.outerWithCloseNotifyAndFlush{UserProvidedDecorator:(*metrics.ResponseWriterDelegator)(0xc0e77aa900), InnerCloseNotifierFlusher:responsewriter.outerWithCloseNotifyAndFlush{UserProvidedDecorator:(*filters.auditResponseWriter)(0xc0ef2d8600), InnerCloseNotifierFlusher:responsewriter.outerWithCloseNotifyAndFlush{UserProvidedDecorator:(*httplog.respLogger)(0xc0d0bba2c0), InnerCloseNotifierFlusher:struct { httpsnoop.Unwrapper; http.ResponseWriter; http.Flusher; http.CloseNotifier; http.Pusher }{Unwrapper:(*httpsnoop.rw)(0xc148e971d0), ResponseWriter:(*httpsnoop.rw)(0xc148e971d0), Flusher:(*httpsnoop.rw)(0xc148e971d0), CloseNotifier:(*httpsnoop.rw)(0xc148e971d0), Pusher:(*httpsnoop.rw)(0xc148e971d0)}}}}, encoder:(*versioning.codec)(0xc11b9448c0), memAllocator:(*runtime.Allocator)(0xc0cec163d8)})
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: E0124 06:10:56.892439   86280 timeout.go:142] post-timeout activity - time-elapsed: 10.377525ms, GET "/api/v1/namespaces/istio-system/configmaps/istio-leader" result: Header called after Handler finished
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: goroutine 691243071 [running]:
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP.func1.1()
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:110 +0x11b
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: panic({0x59b79e0, 0x6ab0310})
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: /usr/local/go/src/runtime/panic.go:890 +0x262
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: golang.org/x/net/http2.(*responseWriter).Header(0xc12abe8148)
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: vendor/golang.org/x/net/http2/server.go:2870 +0xa8
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp.(*respWriterWrapper).Header(0xc1350eb980)
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: vendor/go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp/wrap.go:73 +0x38
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: github.com/felixge/httpsnoop.(*rw).Header(0xc1349a3450)
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: vendor/github.com/felixge/httpsnoop/wrap_generated_gteq_1.8.go:364 +0xee
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: k8s.io/apiserver/pkg/server/httplog.(*respLogger).Header(0xc1343cf8c0)
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: vendor/k8s.io/apiserver/pkg/server/httplog/httplog.go:284 +0x3f
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: k8s.io/apiserver/pkg/endpoints/filters.(*recorder).AddWarning(0xc1349a34a0, {0x0, 0x0}, {0x676ffc6, 0x7b})
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: vendor/k8s.io/apiserver/pkg/endpoints/filters/warning.go:110 +0x5a6
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: k8s.io/apiserver/pkg/warning.AddWarning({0x6ae4e78, 0xc13f02a390}, {0x0, 0x0}, {0x676ffc6, 0x7b})
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: vendor/k8s.io/apiserver/pkg/warning/context.go:59 +0xd8
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: k8s.io/apiserver/pkg/authentication/token/cache.(*cachedTokenAuthenticator).AuthenticateToken(0xc0007ba370, {0x6ae4e78, 0xc13f02a390}, {0xc03a9c3007, 0x3b5})
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: vendor/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go:139 +0x2a5
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: k8s.io/apiserver/pkg/authentication/request/bearertoken.(*Authenticator).AuthenticateRequest(0xc0001f6420, 0xc13d18fd00)
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: vendor/k8s.io/apiserver/pkg/authentication/request/bearertoken/bearertoken.go:63 +0x339
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: k8s.io/apiserver/pkg/authentication/request/union.(*unionAuthRequestHandler).AuthenticateRequest(0xc0004f01e0, 0xc13d18fd00)
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: vendor/k8s.io/apiserver/pkg/authentication/request/union/union.go:56 +0x118
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: k8s.io/apiserver/pkg/authentication/group.(*AuthenticatedGroupAdder).AuthenticateRequest(0xc0001f6440, 0xc13d18fd00)
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: vendor/k8s.io/apiserver/pkg/authentication/group/authenticated_group_adder.go:40 +0x86
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: k8s.io/apiserver/pkg/authentication/request/union.(*unionAuthRequestHandler).AuthenticateRequest(0xc0004f0220, 0xc13d18fd00)
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: vendor/k8s.io/apiserver/pkg/authentication/request/union/union.go:56 +0x118
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: k8s.io/apiserver/pkg/authentication/request/union.(*unionAuthRequestHandler).AuthenticateRequest(0xc000658740, 0xc13d18fd00)
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: vendor/k8s.io/apiserver/pkg/authentication/request/union/union.go:56 +0x118
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: k8s.io/apiserver/pkg/endpoints/filters.withAuthentication.func1({0x6ae3d50, 0xc1659ea040}, 0xc13d18fd00)
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: vendor/k8s.io/apiserver/pkg/endpoints/filters/authentication.go:66 +0x1b2
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: net/http.HandlerFunc.ServeHTTP(0xc00d7b75e0, {0x6ae3d50, 0xc1659ea040}, 0xc13d18fc00)
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: /usr/local/go/src/net/http/server.go:2122 +0x43
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: k8s.io/apiserver/pkg/endpoints/filterlatency.trackStarted.func1({0x6ae3d50, 0xc1659ea040}, 0xc13d18fc00)
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: vendor/k8s.io/apiserver/pkg/endpoints/filterlatency/filterlatency.go:94 +0x36f
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: net/http.HandlerFunc.ServeHTTP(0xc0187ba870, {0x6ae3d50, 0xc1659ea040}, 0xc13f02d600)
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: /usr/local/go/src/net/http/server.go:2122 +0x43
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP.func1()
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:115 +0xd6
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: created by k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:101 +0x3b9
Jan 24 06:10:56 VMS71884 kube-apiserver[86280]: fatal error: concurrent map iteration and map write

```

### What did you expect to happen?

kube-apiserver does not panic even if calling webhook times out or fails

### How can we reproduce it (as minimally and precisely as possible)?

Deploy a webhook, remove the service corresponding to the webhook, and then create pods in batches to trigger webhook calls

### Kubernetes version

<details>

```console
$ kubectl version
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.2", GitCommit:"7f6f68fdabc4df88cfea2dcf9a19b2b830f1e647", GitTreeState:"clean", BuildDate:"2023-05-17T14:20:07Z", GoVersion:"go1.20.4", Compiler:"gc", Platform:"darwin/arm64"}
Kustomize Version: v5.0.1
Server Version: version.Info{Major:"1", Minor:"28+", GitVersion:"v1.28.3-4-rc1-apiserver-mr", GitCommit:"f8ac9363684f399a102efd1027a7db7a6ffa9f3c", GitTreeState:"clean", BuildDate:"2024-01-18T06:21:10Z", GoVersion:"go1.20.4", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
private cloud
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by songminglong** [comment]
> somewhat related to #120507

The scenarios that trigger panic are similar(the apiserver fails to call webhook), but the log shows that the location of panic is different.

**Comment by jiahuif** [comment]
/assign @dims
For context of [`k8s.io/apiserver/pkg/server/filters`](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/filters/OWNERS)
Thank you.
/triage accepted

**Comment by songminglong** [comment]
again, full stack logs:
[kube-apiserver-panic.txt](https://github.com/kubernetes/kubernetes/files/14062163/kube-apiserver-panic.txt)

**Comment by songminglong** [comment]
I'll try to test the mr: https://github.com/kubernetes/kubernetes/pull/114189

**Comment by tkashem** [comment]
Let me try to explain the panic - both goroutines are trying to write to the `Header` of the `ResponseWriter`

first goroutine: This request has timed out, and consequently the `timeout` filter is trying to send a `StatusGatewayTimeout` response to the caller. This is the client facing goroutine from the timeout filter.

https://github.com/kubernetes/kubernetes/blob/a8a1abc25cad87333840cd7d54be2efaf31a3177/staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go#L248-L261

Note that the call to `WriteHeader` is protected by a mutex, that means that if the other goroutine is using the delegated `Responsewriter` object via the `baseTimeoutWriter` adapter, then this panic should not happen.

second goroutine: let's check the other goroutine, it is also trying to write to the `Header` of the `ResponseWriter` object:

https://github.com/kubernetes/kubernetes/blob/a8a1abc25cad87333840cd7d54be2efaf31a3177/staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go#L138-L140

Now `warning.AddWarning` uses an unprotected version of the ResponseWriter object:
https://github.com/kubernetes/kubernetes/blob/a8a1abc25cad87333840cd7d54be2efaf31a3177/staging/src/k8s.io/apiserver/pkg/endpoints/filters/warning.go#L31-L37

The `WithWarningRecorder` filter precedes the `timeout` filter in the execution order, so the `recorder` in  `recorder := &recorder{writer: w}` is not using the concurrent safe `baseTimeoutWriter` implementation.

This is the order:
https://github.com/kubernetes/kubernetes/blob/a8a1abc25cad87333840cd7d54be2efaf31a3177/staging/src/k8s.io/apiserver/pkg/server/config.go#L944-L955

one quick resolution is to move the `WithWarningRecorder` after the timeout filter: https://github.com/kubernetes/kubernetes/pull/115282, but there be could that needs to write warning before that.
Also note that the panic mentioned in 115282 `Header called after Handler finished` is not as sever, it does not cause the apiserver to crash 

In the long term, i think we would want https://github.com/kubernetes/kubernetes/pull/114189

any thoughts? /cc @liggitt @wojtek-t

**Comment by songminglong** [comment]
I tend to remove the timeout filter, the sooner the better, and I think the apiserver panic is the P0-level bug, directly affecting the availability of the cluster

**Comment by likakuli** [comment]
> both goroutines are trying to write to the `Header` of the `ResponseWriter`

why the stack show `concurrent map iteration and map write` instead of `concurrent map write`?

**Comment by liggitt** [comment]
I opened a minimal WIP fix for the warning writer in https://github.com/kubernetes/kubernetes/pull/123034 since the timeout filter removal is a much bigger change to reason through

**Comment by songminglong** [comment]
> I opened a minimal WIP fix for the warning writer in #123034 since the timeout filter removal is a much bigger change to reason through

thanks a lot, I will try to test and cherry-pick the above minimal WIP fix in #123034

**Comment by liggitt** [comment]
sorry, I completely overlooked https://github.com/kubernetes/kubernetes/pull/115282, I only saw https://github.com/kubernetes/kubernetes/pull/114189 and that looked too big to take safely right now

I prefer https://github.com/kubernetes/kubernetes/pull/115282 to https://github.com/kubernetes/kubernetes/pull/123034

**Comment by mborsz** [comment]
Can we backport the fix to 1.28?

**Comment by liggitt** [comment]
> Can we backport the fix to 1.28?

Why 1.28 specifically? I'm not aware of anything that changed in 1.28 that made this more or less prevalent. If this meets backport criteria, going to all active branches could make sense (1.27+)
